### PR TITLE
Use add(extended) to generate array index offset for AArch64.

### PR DESCRIPTION
**Pull request by XiaohongGong**
We want to find a better way to implement array offset calculation.
The old pattern for an int array member loading is:

     and      x4, x3, #0xffffffff
     lsl      x4, x4, #2
     add      x4, x4, #0x10
     ldr      x0, [x2, x4]

And the optimized pattern is:
       
    add      x1, x2, w3, uxtw #2
    ldr      x0, [x1, #16]

Here is a jmh unit test for int array member loading:

```
private final int func(int[] arr) {
        int ret = 0;
        for (int i=0; i < field0; i++) {
            ret += arr[i];
        }
        return ret;
}
```

And the jmh performance results:

```
without this patch       with this patch       units
    1173501.595               948290.420         us/op
```

Here are the additional issues fixed related to this patch:
1. Finetune MarkBasePointers to get better derived reference marking.
  Currently there is an issue when marking live base pointers. The liveness
of the derived reference will decide whether its base pointer is marked.
If one derived reference is dead at some position, its base pointer will be
removed. So if there are two or more derived references that have the same
base pointer, one of them died will make other derived reference cann't
find its base pointer when building the oop map.
  This patch add a map which will record the base pointer and its derived
references. One base pointer can be removed only when all its derived references
are dead.

2. Set RegisterPriority of basepointers according to the STATE_FLAGS when
building its interval.
  As we all know, the constant value will be re-materialized while not be
spilled to stack when there are not free registers. And its location will be
set to invalid after register allocation. So does the constant reference.
So there will be some exceptions when building the oop map, because the compiler
need to save the locations of live references. If a derived reference has a
constant base pointer which is re-materialized, the compiler will not get its
location. So the jvmci will treat the derived reference as a normal reference
because its base part is null, which may cause the gc crash if the derived
reference is not an real object.
  To resolve this issue, I set the register priority of base pointers to "ShouleHaveRegister",
thus the compiler will not materialize the constant base pointer when doing register
allocation. Also the LIRFrameState has a STATE_FLAG for state values, that is
either a register or a stack slot.

3. Fix an issue on ConstantLoadOptimization. When a constant value is replaced with a
new one, we should also check whether this value is a base pointer of a derived reference.
If so, the base pointer of the derived reference should also be replaced with the new value.
Or there will be some issues when doing register allocation.

4. Fix an issue in MatchProcessor, so that we can process platform specific
node as a matchable node.

mx bootstrap passes.

Change-Id: Ia34f2bf044ca6863be075cf000b29524ca44c42d

### CODE REVIEW

**Comment by XiaohongGong** [comment]
Hi there,
Can someone please take a look at this patch? Thank you!

**Comment by sanzinger** [comment]
Hi, thank you for the contribution. Sorry for the delayed review. I have not looked into the code in detail right now but I ran tests on it and it seems to create an issue during register allocation:
```
mx gate -o -t 'CTW:hosted'
```
Gives following error:
```
CompileTheWorld : Waiting for 164293 compiles
[thread:32] scope: CompileTheWorld-32
    [thread:32] scope: CompileTheWorld-32.Compiling.GraalCompiler
    Context: StructuredGraph:27317{HotSpotMethod<ForkJoinTask.expungeStaleExceptions()>}
                [thread:32] scope: CompileTheWorld-32.Compiling.GraalCompiler.BackEnd.EmitLIR.LIRStages.AllocationStage.LinearScan.SSALinearScanLifetimeAnalysis
                Exception raised in scope CompileTheWorld-32.Compiling.GraalCompiler.BackEnd.EmitLIR.LIRStages.AllocationStage.LinearScan.SSALinearScanLifetimeAnalysis: org.graalvm.compiler.debug.GraalError: liveIn set of first block must be empty: {72}
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanLifetimeAnalysisPhase.computeGlobalLiveSets(LinearScanLifetimeAnalysisPhase.java:392)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanLifetimeAnalysisPhase.run(LinearScanLifetimeAnalysisPhase.java:92)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanAllocationPhase.apply(LinearScanAllocationPhase.java:53)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanAllocationPhase.apply(LinearScanAllocationPhase.java:46)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScan.allocate(LinearScan.java:691)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanPhase.run(LinearScanPhase.java:42)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.alloc.lsra.LinearScanPhase.run(LinearScanPhase.java:35)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:117)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:109)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhaseSuite.run(LIRPhaseSuite.java:98)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:117)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:109)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLowLevel(LIRCompilerBackend.java:175)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR0(LIRCompilerBackend.java:155)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR(LIRCompilerBackend.java:113)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitBackEnd(LIRCompilerBackend.java:83)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.target.Backend.emitBackEnd(Backend.java:280)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:145)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compileGraph(GraalCompiler.java:129)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compileHelper(HotSpotGraalCompiler.java:211)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.HotSpotGraalCompiler.compile(HotSpotGraalCompiler.java:229)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilationTask$HotSpotCompilationWrapper.performCompilation(CompilationTask.java:190)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilationTask$HotSpotCompilationWrapper.performCompilation(CompilationTask.java:98)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.CompilationWrapper.run(CompilationWrapper.java:177)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilationTask.runCompilation(CompilationTask.java:344)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.hotspot.CompilationTask.runCompilation(CompilationTask.java:315)
                	at org.graalvm.compiler.hotspot.test.CompileTheWorld.compileMethod(CompileTheWorld.java:982)
                	at org.graalvm.compiler.hotspot.test.CompileTheWorld.access$400(CompileTheWorld.java:117)
                	at org.graalvm.compiler.hotspot.test.CompileTheWorld$1.run(CompileTheWorld.java:919)
                	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
                	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
                	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
                	at java.base/java.lang.Thread.run(Thread.java:834)
                	at jdk.internal.vm.compiler/org.graalvm.compiler.core.CompilerThread.run(CompilerThread.java:42)
                
                Context obj SSALinearScanLifetimeAnalysis
                Context obj org.graalvm.compiler.lir.alloc.lsra.LinearScanPhase@176a9971
                Context obj org.graalvm.compiler.lir.phases.AllocationStage@1e8e1c32
                Context obj org.graalvm.compiler.hotspot.aarch64.AArch64HotSpotNodeLIRBuilder@48dad95d
                Context obj org.graalvm.compiler.hotspot.HotSpotLIRGenerationResult@5b6aaac4
                Context obj org.graalvm.compiler.lir.LIR@180801c9
                Context obj org.graalvm.compiler.nodes.StructuredGraph$ScheduleResult@55e7fa8c
```

**Comment by XiaohongGong** [comment]
@sanzinger  Thank you! I will look at this issue.

**Comment by XiaohongGong** [comment]
@sanzinger I resolved the issue you mentioned above. Also I run the entire CTW task by running the command you gave me, and it has passed. The latest patch has been updated. Please help to take a review. Thank you!

**Comment by XiaohongGong** [comment]
Hello @sanzinger,
  The travis-ci for this PR failed at https://travis-ci.org/oracle/graal/jobs/549533881 when doing the "image demo" task for substratevm. I run the same command `mx --primary-suite-path /home/xiagon01/code/graal/graal/substratevm --java-home=/home/xiagon01/openjdk1.8.0_212-jvmci-20-b03 gate --strict-mode --tags build,test,helloworld` locally on my x86 server. And all the tasks passed:

`Gate task times:
  0:00:00.264424        Versions
  0:00:00.000317        JDKReleaseInfo
  0:00:00.819123        VerifyMultiReleaseProjects
  0:00:01.105065        Clean
  0:00:45.495237        BuildWithJavac
  0:02:47.084437        image demos
  0:00:42.547766        native unittests
`

  I tried to push the patch several times without any changes to retrigger the travis-ci. But it still fails at substratevm verification. So do you know what happens?

Thanks,
XiaohongGong

**Comment by XiaohongGong** [comment]
Hi @sanzinger I rebased this PR and resolved the conflict. Could you please take a look at it? Thank you!

**Comment by sanzinger** [comment]
Hi @XiaohongGong, this issue is quite old now. However, we are still interested into it.

But I can give you an update on the status on this: It is a long time ago when I reviewed this PR. I am not too much familiar with the register allocation. I had a chat with @zapster and we were not sure if this is the right solution. 

I'll try to have a look once again soon. This change will not be part of the 19.3 release. But the next one.

**Comment by XiaohongGong** [comment]
Hi @sanzinger , thank you so much for looking it again! Currently there is an issue this patch that makes several unit tests fails on X86_64 with jdk11 (travis-ci 6377.8, 6377.9). I have fixed it and the new patch is in internal review. Also I will continue to do more tests for it. Thank you!

**Comment by XiaohongGong** [comment]
Hi @sanzinger, I have updated my patch, and please help to take a look at it. 

I resolved the issue I mentioned that several unit tests fails on X86 when build with jdk11 in the new patch.  The old code assumes the derived reference and its base pointer has existed before it is removed, and the new X86 codes hit the assumption. It fails when removing an "Temp" annotated derived-reference value from the base pointer map when it is empty. 

Basically "Temp" value is like "Def" value, which is defined by the instruction. And when doing the liveness analysis, the value defined by one instruction should not be alive. But in the issue code, actually the "Temp" value is not defined by the instruction. They are from the inputs, and will not be used after the instruction. So it cannot guarantee that its base pointer information has been recorded before it's removed.

**Comment by dougxc** [comment]
@sanzinger can this be merged now?

**Comment by sanzinger** [comment]
I can start a test iteration and a review internally. I am not the expert in the register allocation though.

**Comment by sanzinger** [comment]
@zapster: Can you have a look on the register allocator changes please?

**Comment by tkrodriguez** [comment]
The comment on that changeset seems uselessly long.  It seems to be the entire first comment from this PR.  It really should just be a single sentence.

**Comment by sanzinger** [comment]
There are two issues found during testing:

Code formatting:
```
--- a/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java
+++ b/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java
@@ -214,10 +214,10 @@
                 AllocatableValue y = gen.asAllocatable(operand(zeroExtend.getValue()));
                 AllocatableValue baseReference = LIRKind.derivedBaseFromValue(x);
                 LIRKind kind = LIRKind.combineDerived(gen.getLIRKind(addP.stamp(NodeView.DEFAULT)),
-                        baseReference, null);
+                                baseReference, null);
                 Variable result = gen.newVariable(kind);
                 gen.append(new AArch64ArithmeticOp.ExtendedAddShiftOp(result, x, moveSp(y),
-                        extendType, shiftNum));
+                                extendType, shiftNum));
                 return result;
             };
         }
```

and substratevm cannot work with derived references:
```
[helloworld:4000]    classlist:   3,476.00 ms
[helloworld:4000]        (cap):   1,093.22 ms
[helloworld:4000]        setup:   3,528.69 ms
[helloworld:4000]   (typeflow):   6,371.09 ms
[helloworld:4000]    (objects):   4,453.44 ms
[helloworld:4000]   (features):     532.81 ms
[helloworld:4000]     analysis:  11,749.79 ms
[helloworld:4000]     (clinit):     276.92 ms
[helloworld:4000]     universe:     872.20 ms
[helloworld:4000]      (parse):     943.12 ms
[helloworld:4000]     (inline):     817.77 ms
[helloworld:4000]    (compile):   1,756.83 ms
[helloworld:4000]      compile:   3,691.71 ms
Fatal error: org.graalvm.compiler.debug.GraalError: jdk.vm.ci.common.JVMCIError: should not reach here: derived references not supported yet on Substrate VM
	at method: Object java.util.concurrent.ConcurrentHashMap.putVal(Object, Object, boolean)  [Direct call from Object Properties.put(Object, Object)]
	at com.oracle.svm.hosted.code.CompileQueue.defaultCompileFunction(CompileQueue.java:989)
	at com.oracle.svm.hosted.code.CompileQueue.doCompile(CompileQueue.java:920)
	at com.oracle.svm.hosted.code.CompileQueue$CompileTask.run(CompileQueue.java:259)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:171)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: jdk.vm.ci.common.JVMCIError: should not reach here: derived references not supported yet on Substrate VM
	at jdk.internal.vm.ci/jdk.vm.ci.common.JVMCIError.shouldNotReachHere(JVMCIError.java:48)
	at com.oracle.svm.core.heap.SubstrateReferenceMapBuilder.addLiveValue(SubstrateReferenceMapBuilder.java:62)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.RegStackValueSet$1.visitValue(RegStackValueSet.java:155)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.ValueConsumer.visitValue(ValueConsumer.java:51)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.util.IndexedValueMap.visitEach(IndexedValueMap.java:143)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.RegStackValueSet.addLiveValues(RegStackValueSet.java:159)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarkerPhase$Marker.processState(LocationMarkerPhase.java:97)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarkerPhase$Marker.processState(LocationMarkerPhase.java:57)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarker$1.doState(LocationMarker.java:160)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.LIRInstructionClass.forEachState(LIRInstructionClass.java:300)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.LIRInstruction.forEachState(LIRInstruction.java:294)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarker.processInstructionBottomUp(LocationMarker.java:151)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarker.processBlock(LocationMarker.java:113)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarker.build(LocationMarker.java:83)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarkerPhase.run(LocationMarkerPhase.java:54)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.dfa.LocationMarkerPhase.run(LocationMarkerPhase.java:50)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:117)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:109)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhaseSuite.run(LIRPhaseSuite.java:98)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:117)
	at jdk.internal.vm.compiler/org.graalvm.compiler.lir.phases.LIRPhase.apply(LIRPhase.java:109)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLowLevel(LIRCompilerBackend.java:180)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR0(LIRCompilerBackend.java:160)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitLIR(LIRCompilerBackend.java:113)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.gen.LIRCompilerBackend.emitBackEnd(LIRCompilerBackend.java:83)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.target.Backend.emitBackEnd(Backend.java:279)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:145)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compileGraph(GraalCompiler.java:129)
	at com.oracle.svm.hosted.code.CompileQueue.defaultCompileFunction(CompileQueue.java:958)
	... 9 more
Error: Image build request failed with exit status 1
```

You can run it with:
```
cd substratevm
mx build
javac HelloWorld.java
mx native-image HelloWorld
```

**Comment by XiaohongGong** [comment]
> substratevm cannot work with derived references

Hi @sanzinger , do you know whether it has plan to support the derived reference for SVM in future? And because the graal compiler might actually generate a derived reference, how will SVM handle such values for GC in future?

Thanks,
Xiaohong

**Comment by XiaohongGong** [comment]
Can we add a flag to tell the compiler whether "derived reference" is supported? And enable this optimization if it is supported, and disable if it is not supported.

**Comment by XiaohongGong** [comment]
@sanzinger I resolved the two issues and have updated my patch. For the derived reference issue in SVM, I just added a flag to disable this optimization on SVM. Could you please take a look at it again? Thank you so much!

**Comment by sanzinger** [comment]
@XiaohongGong I have merged the MarkBasePointersPhase this afternoon. I also have adapted this PR accordingly. Please have a look, if it is okay, we can merge this change.

**Comment by XiaohongGong** [comment]
Hi @sanzinger ,

Thank you so much for your updating! It looks like good for me.  And also all our internal tests pass without any issue. So I think it's time to merge this PR. 

Thanks,
Xiaohong
